lib/types.nix

```
@@ -629,7 +629,7 @@ rec {
     unique = { message }: type: mkOptionType rec {
       name = "unique";
       inherit (type) description descriptionClass check;
-      merge = mergeUniqueOption { inherit message; };
+      merge = mergeUniqueOption { inherit message; inherit (type) merge; };
```

This feels like a small layer violation to me, `lib/options.nix` should only _provide` merge functions, not make use of them.

Instead we could do this:
```suggestion
      merge =
        if length defs == 1 then
          # Makes sure that nested values are also type-checked
          type.merge loc defs
        else
          throw "The option `${showOption loc}' is defined multiple times while it's expected to be unique.\n${message}\nDefinition values:${showDefs defs}\n${prioritySuggestion}";
```

---

I don't think there was any real layering to begin with. Options _use_ types, but types _use_ `options.nix` to implement merging.

Also I find it surprising that you object to the construction of merge functions from other merge functions.
Constructing \<x> from other \<x> is exactly what makes languages and libraries scale.

---

Fair, just felt a bit like that, but I don't really mind that. Though I do like the simplicity of the suggestion still :)